gapic/src/main/com/google/gapid/perfetto/models/PowerSummaryTrack.java

```diff
@@ -0,0 +1,184 @@
+            .collect(Collectors.toList());
+
+    if (powerRails.size() == 0) {
+      return data;
+    }
+
+    // Power Group
+    viewSql = buildViewSql(powerRails);
+    PowerSummaryTrack powerSummaryTrack = new PowerSummaryTrack(data.qe, powerRails.size());
+
+    // Power Rails tracks.
+    for (CounterInfo powerRail : powerRails) {
+      unit = powerRail.unit;
+      CounterTrack powerRailTrack = new CounterTrack(data.qe, powerRail);
+      data.tracks.addTrack(
+          powerSummaryTrack.getId(),
+          powerRailTrack.getId(),
+          powerRail.name,
+          single(
+              state -> new CounterPanel(state, powerRailTrack, POWER_RAIL_COUNTER_TRACK_HEIGHT),
+              true,
+              true));
+    }
+
+    Group.UiFactory ui = group(state -> new PowerSummaryPanel(state, powerSummaryTrack), false);
+    data.tracks.addLabelGroup(null, powerSummaryTrack.getId(), "Power Usage", ui);
+    return data;
+  }
+
+  private static String buildViewSql(List<CounterInfo> powerRails) {
+    minValue = powerRails.get(0).min;
+    maxValue = powerRails.get(0).max;
+    StringBuilder sb =
+        new StringBuilder()
+            .append(
+                "select aggregateTable.ts ts, sum(aggregateTable.value) totalValue,"
+                    + " aggregateTable.dur dur from")
+            .append(" (")
+            .append(VIEW_SQL_MONOTONIC)
+            .append(powerRails.get(0).id)
+            .append(" window win as (order by ts) ")
+            .append(") as aggregateTable");
+    for (int i = 1; i < numPowerRailTracks; i++) {
+      sb.append("left join (")
+          .append(VIEW_SQL_MONOTONIC)
+          .append(powerRails.get(i).id)
+          .append(" window win as (order by ts) ")
+          .append(") ");
+      minValue = Math.min(minValue, powerRails.get(i).min);
+      maxValue = Math.max(maxValue, powerRails.get(i).max);
+    }
+    sb.append(" group by ts");
+    minValue = Math.min(0, minValue);
+    return sb.toString();
+  }
+
+  @Override
+  protected ListenableFuture<?> initialize() {
+    String vals = tableName("vals");
+    String span = tableName("span");
+    String window = tableName("window");
+    return qe.queries(
+        dropTable(span),
+        dropTable(window),
+        dropView(vals),
+        createView(vals, viewSql),
+        createWindow(window),
+        createSpan(span, vals + ", " + window));
+  }
+
+  @Override
+  protected ListenableFuture<Data> computeData(DataRequest req) {
+    Window win = Window.quantized(req, 5);
+    return transformAsync(win.update(qe, tableName("window")), $ -> computeData(req, win));
+  }
+
+  private ListenableFuture<Data> computeData(DataRequest req, Window win) {
+    return transform(
+        qe.query(sql()),
+        res -> {
+          int rows = res.getNumRows();
+          if (rows == 0) {
+            return Data.empty(req);
+          }
+
+          Data data = new Data(req, new long[rows + 1], new double[rows + 1]);
```

Comment:
Why is there a `rows + 1` here? Would prefer comment clarifying magic numbers (and the `rows -1` below

Reply:
Makes sense.